### PR TITLE
feat: Add Google OR-Tools CP-SAT solver

### PR DESCRIPTION
**Pull request by gabriel-gehrke**
This PR adds a bridge to the **Google OR-Tools CP-SAT solver** via the [`cp_sat`](https://crates.io/crates/cp_sat) Rust crate (v0.4.0), accessible through the `cp_sat` feature flag.

CP-SAT is one of the fastest open-source solvers for combinatorial (integer) optimization problems. It supports integer and boolean variables, linear objectives, linear constraints, and advanced features like solution hints (warm-starting), time limits, MIP gaps, parallel search with different search strategies, and verbose logging.

## Implementation

The bridge lives in `src/solvers/cp_sat.rs` and covers:

- **`cp_sat()` function**: Converts `UnsolvedProblem` to `CpSatProblem`
- **`SolverModel` impl**: `add_constraint()` with equality/inequality support, `solve()` with full status mapping
- **`WithInitialSolution` impl**: Warm-start hints via `CpModelBuilder::add_hint()`
- **`WithTimeLimit` impl**: Set `max_time_in_seconds` on internal parameters
- **`WithMipGap` impl**: Set `relative_gap_limit` on internal parameters
- **Additional API**: configuring verbosity and some search parameters (`set_verbose()`, `set_log_to_response()`, `set_num_search_workers()`, `set_random_seed()`, ...)
- **`CpSatSolution`** wraps `CpSolverResponse` with access to `response_stats()`, `solution_log()`, `additional_solutions()`, `into_response()`
- **24 unit tests** covering core functionality, edge cases, and error handling (all passing)

## Limitations & Rough Edges

Unlike most other good_lp solvers, the `cp_sat` crate links dynamically to a very heavy OR-Tools shared library. This is handled by the cp_sat crate but it does so a little clumsily, relying on correct setting of environment variables (Note I am already working on a better ortools to rust interface, which could replace this crate in the future). Currently, users must:
   - Install OR-Tools on their system (prebuilt binaries provided for different Linux distribution, macos and WIndows via Github releases)
   - Set `ORTOOLS_PREFIX`, `RUSTFLAGS="-L /opt/ortools/lib -lprotobuf"`, `LD_LIBRARY_PATH=/opt/ortools/lib`

As the OR-Tools prebuilt binary is platform-specific, I currently pin it to `ubuntu-24.04` in CI. The CI job runs only cp_sat-specific lib tests (`cargo test --lib -- cp_sat`), not integration tests which use continuous variables and would panic.

For now I did not put it into `all_default_solvers` due to the heavy external dependency (OR-Tools shared libraries), cp_sat is opt-in only, similar to `cplex-rs`.

## Progress on Fixes

- [x] No more rounding. Reject non-integer values instead (turns model invalid, clean error is returned during `solve()`.
- [x] Don't special-case CI workflow
- [x] Replace lambda with imperative logic
- [x] Don't require manual env vars
- [x] Point to upstream OR-Tools install docs
- [x] Remove redundant `dep:cp_sat`
- [x] Remove duplicate solver list in `src/lib.rs`

### CODE REVIEW

File: src/solvers/cp_sat.rs [hunk 1]
```diff
@@ -0,0 +1,850 @@
...
+) -> IntVar {
+    let name = if def.name.is_empty() {
+        format!("v{}", var.index())
+    } else {
+        def.name.clone()
+    };
+
+    // Translate bounds to CP-SAT domain
+    let domain_lower = if def.min.is_finite() {
+        round_i64(def.min)
```
Comment (TeXitoi) [inline src/solvers/cp_sat.rs]: rounding is not correct. lower must be ceiled, and upper must be floored.

File: src/solvers/cp_sat.rs [hunk 2]
```diff
@@ -0,0 +1,850 @@
...
+    pub fn as_inner(&self) -> &CpModelBuilder {
+        &self.model
+    }
+
+    /// Builds a `LinearExpr` from a good_lp `Constraint`'s expression,
+    /// converting f64 coefficients to i64.
+    fn expr_from_constraint_expression(&self, expression: &crate::Expression) -> LinearExpr {
+        let mut linear = LinearExpr::default();
+        for (var, coeff) in expression.linear.coefficients.iter() {
+            let coeff_i64 = round_i64(*coeff);
```
Comment (TeXitoi) [inline src/solvers/cp_sat.rs]: I’m afraid these rounding will be very error prone for the users.

File: .github/workflows/rust.yml [hunk 3]
```diff
@@ -150,17 +150,38 @@ jobs:
...
+        env:
+          ORTOOLS_PREFIX: /opt/ortools
+          RUSTFLAGS: -L /opt/ortools/lib -lprotobuf
+          LD_LIBRARY_PATH: /opt/ortools/lib
+      - name: Run tests with cp_sat
+        run: cargo test --no-default-features --features cp_sat --lib -- cp_sat
+        env:
+          ORTOOLS_PREFIX: /opt/ortools
+          RUSTFLAGS: -L /opt/ortools/lib -lprotobuf
+          LD_LIBRARY_PATH: /opt/ortools/lib
```
Comment (lovasoa) [inline .github/workflows/rust.yml]: can we avoid special-casing the new solver like that ? Use the standard system library path, standard build.rs in the underlying crate, etc. Ideally you would add an option to statically link the c++ deps in the underlying crate and use it here. The goal of good-lp is to be as user-friendly and zero-config as possible.

File: src/solvers/cp_sat.rs [hunk 4]
```diff
@@ -0,0 +1,850 @@
...
+                "CP-SAT does not support continuous variables. \
+                 Variable {} (index {}) was not declared as integer or binary. \
+                 Please use `.integer()` or `.binary()` on the variable definition.",
+                if def.name.is_empty() {
+                    format!("v{}", var.index())
+                } else {
+                    def.name.clone()
+                },
+                var.index()
+            );
```
Comment (lovasoa) [inline src/solvers/cp_sat.rs]: can we avoid crashing the entire program here ? Just let the model build and return an error on solve

**Review by TeXitoi — COMMENTED** [review]
This "integer only model" makes cp_sat difficult to use as a LP solver. Ideally, the weights would be scaled, but that’s not really doable without a global analysis of the LP.

File: src/solvers/cp_sat.rs [hunk 5]
```diff
@@ -0,0 +1,850 @@
...
+    }
+
+    problem
+}
+
+/// Rounds `f64` to `i64` with saturation at the numeric limits.
+/// Panics on NaN.
+fn round_i64(x: f64) -> i64 {
+    if x.is_nan() {
+        panic!("CP-SAT received NaN");
```
Comment (lovasoa) [inline src/solvers/cp_sat.rs]: please do not crash the user's program. Handle errors cleanly

File: src/solvers/cp_sat.rs [hunk 6]
```diff
@@ -0,0 +1,850 @@
...
+///
+/// Bound mapping: `min`/`max` from `VariableDefinition` to CP-SAT domain interval `(lo, hi)`.
+/// Non-finite bounds (infinity) map to `i64::MIN` / `i64::MAX`.
+fn create_cp_sat_var(
+    model: &mut CpModelBuilder,
+    def: &VariableDefinition,
+    var: Variable,
+) -> IntVar {
+    let name = if def.name.is_empty() {
+        format!("v{}", var.index())
```
Comment (lovasoa) [inline src/solvers/cp_sat.rs]: is this mandatory ? This allocates one string per variable. Is there a way to solve a problem without allocating all these tiny strings ?

File: src/solvers/cp_sat.rs [hunk 7]
```diff
@@ -0,0 +1,850 @@
...
+    };
+
+    // Validate: lower bound must not exceed upper bound
+    assert!(
+        domain_lower <= domain_upper,
+        "Invalid variable bounds: lower={} > upper={} for variable '{}'",
+        def.min,
+        def.max,
+        name
+    );
```
Comment (lovasoa) [inline src/solvers/cp_sat.rs]: do not crash

File: src/solvers/cp_sat.rs [hunk 8]
```diff
@@ -0,0 +1,850 @@
...
+    fn solve(self) -> Result<Self::Solution, Self::Error> {
+        let response = self.model.solve_with_parameters(&self.params);
+
+        let has_time_limit = self.params.max_deterministic_time.is_some()
+            || self.params.max_time_in_seconds.is_some();
+        let has_gap_limit =
+            self.params.relative_gap_limit.is_some() || self.params.absolute_gap_limit.is_some();
+
+        let status = response.status();
+        let status_from_limits = || {
```
Comment (lovasoa) [inline src/solvers/cp_sat.rs]: why does this need to be a function?

**Comment by gabriel-gehrke** [comment]
> This "integer only model" makes cp_sat difficult to use as a LP solver. Ideally, the weights would be scaled, but that’s not really doable without a global analysis of the LP.

The [CP-SAT primer](https://d-krupke.github.io/cpsat-primer/04_modelling.html?highlight=fractional#linear-constraints) by Dominik Krupke highlights this about fractional (continuous values):


>Let us look at the following example with two linear equality constraints:
>
> x−y=0
> 4−x=2y
>
> x,y≥0
>
> You can verify that x=4/3 and y=4/3 is a feasible solution. However, coding this in CP-SAT results in an infeasible solution:
>
> model = cp_model.CpModel()
> x = model.new_int_var(-100, 100, "x")
> y = model.new_int_var(-100, 100, "y")
>
> model.add(x - y == 0)
> model.add(4 - x == 2 * y)
>
> solver = cp_model.CpSolver()
> status = solver.solve(model)
> assert status == cp_model.INFEASIBLE
>
> Even using scaling techniques, such as multiplying integer variables by 1,000,000 to increase the resolution, would not render the model feasible. While common linear programming solvers would handle this model without issue, CP-SAT struggles unless modifications are made to eliminate fractions, such as multiplying all terms by 3. However, this requires manual intervention, which undermines the idea of using a solver. These limitations are important to consider, although such scenarios are rare in practical applications.<


This shows that automatic scaling can not reliably cover all buliding blocks that good_lp offers when using CP-SAT. Personally, I would be in favor of leaving this to the user.

**Comment by gabriel-gehrke** [comment]
Maybe we could add a warning whenever a fractional value is passed to the rounding method, informing the user that their value will be rounded.

Even considering the limitations, I think it would be highly beneficial to support CP-SAT in good_lp. It is very performant, and there are many combinatorial problem classes that do not necessarily involve continuous variables (covering, routing, knapsack packing, vehicle routing, etc.). Having CP-SAT alongside the LP and MIP solvers would also allow combining LP for relaxations and switching to CP-SAT for the integer solving phase.

**Comment by TeXitoi** [comment]
Note that my remark is not limited to linear variables, but also fractional weights in the objective and constraint. If the user says `x + 0.5y > 1.5`, it will not do the intended thing. In this case, multiplying by 2 all the weight will work, but more complex cases exists. As long as you don’t use fractional variables, you can scale at the constraint/objective level. You even have a factor to "correct" the scaling of the objective.

File: README.md [hunk 9]
```diff
@@ -210,6 +211,36 @@ trait, which allows you to access the dual values of the constraints (the shadow
...
+cd /opt/ortools && sudo make test
+```
+
+Then set the following environment variables when building and running:
+
+```bash
+export ORTOOLS_PREFIX=/opt/ortools
+export RUSTFLAGS="-L /opt/ortools/lib -lprotobuf"
+export LD_LIBRARY_PATH=/opt/ortools/lib
+```
```
Comment (lovasoa) [inline README.md]: that's non standard, and painful for users

File: README.md [hunk 10]
```diff
@@ -210,6 +211,36 @@ trait, which allows you to access the dual values of the constraints (the shadow
...
+through its C++ API.
+
+To use CP-SAT, you must install the OR-Tools shared library on your system.
+On Ubuntu 24.04, you can install it as follows:
+
+```bash
+curl -LO 'https://github.com/google/or-tools/releases/download/v9.15/or-tools_amd64_ubuntu-24.04_cpp_v9.15.6755.tar.gz'
+sudo mkdir -p /opt/ortools
+sudo tar -xzf or-tools_amd64_ubuntu-24.04_cpp_v9.15.6755.tar.gz -C /opt/ortools --strip-components=1
+cd /opt/ortools && sudo make test
```
Comment (lovasoa) [inline README.md]: Maybe we should point to upstream instructions

File: Cargo.toml [hunk 11]
```diff
@@ -22,6 +22,7 @@ default = ["coin_cbc", "singlethread-cbc"]
 singlethread-cbc = ["coin_cbc?/singlethread-cbc"]
 scip = ["russcip"]
 scip_bundled = ["russcip?/bundled"]
+cp_sat = ["dep:cp_sat"]
```
Comment (lovasoa) [inline Cargo.toml]: is that needed ?

File: src/lib.rs [hunk 12]
```diff
@@ -3,6 +3,19 @@
...
+//!
+//! - [`coin_cbc`](solvers::coin_cbc) - [COIN-OR CBC](https://www.coin-or.org/Cbc/) (enabled by default)
+//! - [`cp_sat`](solvers::cp_sat) - [Google OR-Tools CP-SAT](https://developers.google.com/optimization/cp/cp_solver) (requires `cp_sat` feature)
+//! - [`microlp`](solvers::microlp) - A lightweight LP solver
+//! - [`lpsolve`](solvers::lpsolve) - [LpSolve](http://lpsolve.sourceforge.net/)
+//! - [`highs`](solvers::highs) - [HiGHS](https://highs.dev/)
+//! - [`scip`](solvers::scip) - [SCIP](https://scipopt.org/)
+//! - [`lp-solvers`](solvers::lp_solvers) - A wrapper around several LP solvers
+//! - [`clarabel`](solvers::clarabel) - [Clarabel](https://github.com/oxfordcontrol/Clarabel.rs)
+//!
```
Comment (lovasoa) [inline src/lib.rs]: we can inline the readme if you want. But let's not maintain two lists

**Review by lovasoa — CHANGES_REQUESTED** [review]
thanks ! that's a good addition.

See my suggestions below.

I haven't checked the quality of the underlying crate itself, I trust you with it.

**Comment by lovasoa** [comment]
@gabriel-gehrke : also from the test logs, I see it prints to stdout. Can we avoid that ?

**Comment by lovasoa** [comment]
For non integer weights : can we just reject them with a clean error ? And document that solver accepts only integers. Let users do the rounding on their side

**Comment by gabriel-gehrke** [comment]
> @gabriel-gehrke : also from the test logs, I see it prints to stdout. Can we avoid that ?

I implemented a configurable verbosity flag similar to how Highs can print to stdout (so CP-SAT just does that, for example for debugging). It can also that provide the log after terminating. In one of the tests it basically tests that this flag can be set. I can remove that test, as it is more like a "this does not cause an error" thing, and not checking whether it actually prints to stdout.

**Comment by gabriel-gehrke** [comment]
Thank you for the detailed feedback, I will make sure to fix these issues.

**Comment by lovasoa** [comment]
> In one of the tests it basically tests that this flag can be set.

Oh ok no problem then, you can leave that part as is. Sorry!

File: .github/workflows/rust.yml (hunk 3 above)
Comment (gabriel-gehrke) [reply .github/workflows/rust.yml]: > Use the standard system library path, standard build.rs in the underlying crate, etc. Ideally you would add an option to statically link the c++ deps in the underlying crate and use it here. The goal of good-lp is to be as user-friendly and zero-config as possible

Two problems here:
1) The ortools C++ library is quite heavy and it has a lot of dependencies: Abseil, Protobuf, zlib, pthreads, other solvers again, ... I have not looked into whether that is possible at all yet, but my instinct says: NO. It also takes a non-trivial time to build from source, in the order of a couple of minutes even on a 24 thread cpu. So I think most users will just install this library as a binary release.
2) The underlying crate assumes the ortools library to lie at "/opt/ortools", so that is actually the default and not the standard system library path. Another location can be provided by setting ORTOOLS_PREFIX. I agree that we should simplify this. And this is actually one of the reasons why I am working on my own ortools bridge, which will not only provide an interface to CP-SAT but also to some of their other optimizers, and handle linking better via CMake.

File: src/solvers/cp_sat.rs (hunk 4 above)
Comment (gabriel-gehrke) [reply src/solvers/cp_sat.rs]: This confuses me because I found analogous behavior in the clarabel solver interface:

```rust
/// The [clarabel](https://oxfordcontrol.github.io/ClarabelDocs/stable/) solver,
/// to be used with [UnsolvedProblem::using].
pub fn clarabel(to_solve: UnsolvedProblem) -> ClarabelProblem {
    
    // [...]

    for (var, def) in variables.iter_variables_with_def() {
        if def.is_integer {
            panic!("Clarabel doesn't support integer variables")
        }
        // [...]
    }
    p
}
```

File: src/solvers/cp_sat.rs (hunk 6 above)
Comment (gabriel-gehrke) [reply src/solvers/cp_sat.rs]: Yes, there is. There actually is unnamed variable creation, I just switched to that now in case no name is given.

File: Cargo.toml (hunk 11 above)
Comment (gabriel-gehrke) [reply Cargo.toml]: I looked into this and as far as I understand, it's not strictly necessary. Cargo automatically creates a feature with the same name as any optional dependency, so the `cp_sat` feature would exist even without the explicit `cp_sat = ["dep:cp_sat"]` line.

I think I added this because I tried adding cp_sat = ["cp_sat"] before and that failed, so I looked for solutions. But can probably be dropped.  Only benefit is the `dep:` prefix makes the relationship explicit rather than relying on the implicit auto-generated feature.

File: src/solvers/cp_sat.rs (hunk 4 above)
Comment (lovasoa) [reply src/solvers/cp_sat.rs]: yeah, clarabel should not do that either ! pr welcome !

File: Cargo.toml (hunk 11 above)
Comment (lovasoa) [reply Cargo.toml]: yes, you can drop it

**Comment by matth2k** [comment]
No contribution on my end here, but this would be a really cool backend to have for what I'm working on. cp-sat is a speedy solver for the problems I used it on.

**Comment by gabriel-gehrke** [comment]
So I've been thinking about the whole approach of 'not crashing the user's program even if inputs are invalid'. Due to the trait signatures in this crate (notably `SolverModel::add_constraint` returning a `ConstraintReference`, and the `Solver` impl using `FnMut(UnsolvedProblem) -> CpSatProblem`), pulling this off without becoming a pain in the ass is surprisingly non-trivial.

My proposal:

- Store state logic inside `CpSatProblem` via a private `ProblemState` enum that represents either a `Valid` state (with `CpModelBuilder` + variable vector) or an `Invalid` state (with just a `ResolutionError`)
- While in `Valid` state, add variables and constraints to the model normally
- As soon as an invalid definition is encountered (continuous variable, lower > upper, NaN coefficient), transition to `Invalid` and immediately fail in `solve()`, which is the earliest point in the model-creation + solving process that returns a `Result`

About variable/constraint handles:

- `add_constraint()` in `Invalid` state could return a constant `ConstraintReference { index: 0 }`. Since `solve()` will return `Err`, the user can never meaningfully use that reference, but code that depens on these being unique could behave unexpectedly.

I would appreciate feedback on this approach

**Comment by lovasoa** [comment]
@gabriel-gehrke : yes, I think that's the way it should work. A Valid(InnerModel)/Invalid enum, all methods implemented on InnerModel, the top level implementation just forwards calls to the inner struct or does nothing.

File: src/solvers/cp_sat.rs [hunk 13]
```diff
@@ -0,0 +1,1152 @@
...
+             or variable bounds. Received `{x}`, which is not an integer. \
+             Please round or scale your values to integers."
+        )));
+    }
+    if !(x >= -(2f64.powi(63)) && x < 2f64.powi(63)) {
+        return Err(ResolutionError::Str(format!(
+            "The CP-SAT solver value `{x}` is out of the i64 range."
+        )));
+    }
+    Ok(x as i64)
```
Comment (lovasoa) [inline src/solvers/cp_sat.rs:134]: we could use try_from, no ?

File: README.md [hunk 14]
```diff
@@ -210,6 +211,36 @@ trait, which allows you to access the dual values of the constraints (the shadow
 
 [clarabel]: https://github.com/oxfordcontrol/Clarabel.rs
 
+### [CP-SAT][cp_sat]
+
+[CP-SAT](https://developers.google.com/optimization/cp/cp_solver) is Google's fast constraint programming solver,
+part of [OR-Tools](https://developers.google.com/optimization). It is one of the fastest open-source solvers
+for combinatorial (integer) optimization problems, supporting both integer and boolean variables.
+CP-SAT does **not** support continuous (floating-point) variables; using a non-integer variable will cause a panic.
```
Comment (lovasoa) [inline README.md:219]: just return an error

**Comment by gabriel-gehrke** [comment]
I opened a [PR on the cp_sat crate](https://github.com/KardinalAI/cp_sat/pull/43). That PR fixes the dependency on 2 of the 3 environment variables my setup currently depends on. I do not have high hopes for it to get merged, unfortunately, as the repository has multiple open pull requests without any reaction from the maintainer(s).

**Comment by lovasoa** [comment]
@TeXitoi , maybe you could have a look ?
